### PR TITLE
GH-40727: [C++][Gandiva] 'ilike' function does not work

### PR DESCRIPTION
**Pull request by lriggs**
### Rationale for this change

Bug fix for GH-40727

### What changes are included in this PR?

Register the 'ilike' function.
Added unit test.

### Are these changes tested?

Yes, unit test.

### Are there any user-facing changes?

No.

* GitHub Issue: #40727

### CODE REVIEW

File: cpp/src/gandiva/tests/utf8_test.cc [hunk 1]
```diff
@@ -221,6 +221,47 @@ TEST_F(TestUtf8, TestLike) {
...
+TEST_F(TestUtf8, TestIlike) {
+  // schema for input fields
+  auto field_a = field("a", utf8());
+  auto schema = arrow::schema({field_a});
+
+  // output fields
+  auto res = field("res", boolean());
+
+  // build expressions.
+  // like(literal(s), a)
```
Comment (kou) [inline cpp/src/gandiva/tests/utf8_test.cc:233]:
```suggestion
  // ilike(literal(s), a)
```

File: cpp/src/gandiva/tests/utf8_test.cc [hunk 2]
```diff
@@ -221,6 +221,47 @@ TEST_F(TestUtf8, TestLike) {
...
+
+  auto node_a = TreeExprBuilder::MakeField(field_a);
+  auto literal_s = TreeExprBuilder::MakeStringLiteral("%spark%");
```
Comment (kou) [inline cpp/src/gandiva/tests/utf8_test.cc:236]: Could you use case mixed string literal?

File: cpp/src/gandiva/tests/utf8_test.cc [hunk 3]
```diff
@@ -221,6 +221,47 @@ TEST_F(TestUtf8, TestLike) {
...
+  auto expr = TreeExprBuilder::MakeExpression(is_like, res);
+
+  // Build a projector for the expressions.
+  std::shared_ptr<Projector> projector;
+  auto status = Projector::Make(schema, {expr}, TestConfiguration(), &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
+
+  // Create a row-batch with some sample data
+  int num_records = 4;
+  auto array_a = MakeArrowArrayUtf8({"park", "sparkle", "bright spark and fire", "spark"},
```
Comment (kou) [inline cpp/src/gandiva/tests/utf8_test.cc:247]: Could you use case mixed test data?

**Comment by lidavidm** [comment]
Can you rebase again? This has picked up unrelated changes

**Comment by lriggs** [comment]
> Can you rebase again? This has picked up unrelated changes

Trying to, not sure how these got in there.

**Review by kou — APPROVED** [review]
+1

**Pull request merged**